plugin/loader/loader.go

```diff
@@ -81,41 +84,83 @@ func (loader *PluginLoader) Initialize() error {
 	return nil
 }
 
-//Run the plugins
-func (loader *PluginLoader) Run() error {
+// Inject hooks all the plugins into the appropriate subsystems.
+func (loader *PluginLoader) Inject() error {
 	for _, pl := range loader.plugins {
-		switch pl := pl.(type) {
-		case plugin.PluginIPLD:
-			err := runIPLDPlugin(pl)
+		if pl, ok := pl.(plugin.PluginIPLD); ok {
+			err := injectIPLDPlugin(pl)
+			if err != nil {
+				return err
+			}
+		}
+		if pl, ok := pl.(plugin.PluginTracer); ok {
+			err := injectTracerPlugin(pl)
 			if err != nil {
 				return err
 			}
-		case plugin.PluginTracer:
-			err := runTracerPlugin(pl)
+		}
+		if pl, ok := pl.(plugin.PluginDatastore); ok {
+			err := injectDatastorePlugin(pl)
 			if err != nil {
 				return err
 			}
-		case plugin.PluginDatastore:
-			err := fsrepo.AddDatastoreConfigHandler(pl.DatastoreTypeName(), pl.DatastoreConfigParser())
+		}
+	}
+	return nil
+}
+
+// Start starts all long-running plugins.
+func (loader *PluginLoader) Start(iface coreiface.CoreAPI) error {
+	for i, pl := range loader.plugins {
+		if pl, ok := pl.(plugin.PluginDaemon); ok {
+			err := pl.Start(iface)
 			if err != nil {
+				closePlugins(loader.plugins[i:])
```

Comment:
I believe this is a bug. It should read `loader.plugins[:i]` as all plugins _before_ `i` have been started not those _after_ `i`.

Reply:
This seems to be the case, do you mind sending a PR?

Reply:
I'll send a PR.